src/mlpack/methods/reinforcement_learning/environment/frozen_lake.hpp

```diff
@@ -0,0 +1,452 @@
+   *    limited to nRows and nCols, respectively.
+   */
+  State InitialSample(std::vector<std::vector<char>> board, size_t nRows, size_t nCols)
+  {
+    stepsPerformed = 0;
+    this->nRows = nRows;
+    this->nCols = nCols;
+    boardDescription = board;
+    return State(nRows, nCols);
+  }
+
+  /**
+   * This function checks if the agent has reached the terminal state.
+   *
+   * @param state The desired state.
+   * @return true if state is a terminal state, otherwise false.
+   */
+  bool IsTerminal(const State& state) const
+  { 
+    if (maxSteps != 0 && stepsPerformed >= maxSteps)
+    {
+      Log::Info << "Episode terminated due to the maximum number of steps"
+          "being taken.\n";
+      return true;
+    }
+    else if (boardDescription[state.CurRow()][state.CurCol()] == 'H')
+    {
+      Log::Info << "Episode terminated due to agent falling in the hole.\n";
+      return true;
+    }
+    else if (boardDescription[state.CurRow()][state.CurCol()] == 'G')
+    {
+      Log::Info << "Episode terminated, agent reached the goal.\n";
+      return true;
+    }
+    return false;
+  }
+
+  //! Get the number of steps performed.
+  size_t StepsPerformed() const { return stepsPerformed; }
+
+  //! Get the maximum number of steps allowed.
+  size_t MaxSteps() const { return maxSteps; }
+
+  //! Set the maximum number of steps allowed.
+  size_t& MaxSteps() { return maxSteps; }
+
+  //! Get the board description.
+  std::vector<std::vector<char>> Description() const {return boardDescription;}
+
+  //! Set the board description.
+  std::vector<std::vector<char>>& Description() {return boardDescription;}
+
+ private:
+  /**
+   * Utility function that helps generate random environment board.
+   * 
+   * @return board, a 2D array of characters that hold the board description.
+   *    'S' is the Start Tile, the starting position of the agent.
+   *    'G' is the Goal Tile, the agent can walk into a goal tile to win.
+   *    'F' is a Frozen Tile, the agent can walk on it.
+   *    'H' is a Hole Tile, the agent can fall into it and the game ends.
+   */
+  std::vector<std::vector<char>> generateRandomBoard()
+  {
+    bool valid = false;
+    std::vector<std::vector<char>> Board;
+    //! TODO: Tentative, implement max step to control whether
+    //    the we can never generate a possible board.
+    while (!valid)
+    {
+      // Randomly choose tiles in the board (discrete random distribution).
+      Board = genBoardHelper(nRows, nCols, platformRate);
+      valid = dfsHelper(Board, nRows, nCols);
+    }
+    return Board;
+  }
+
+  /**
+   * Perform depth-firsth search to see if the environment board has 
+   * a solution or not.
+   * 
+   * @param candidateBoard: the environment board description.
+   * @param nRows: the height of the environment board.
+   * @param nCols: the width of the environment board.
+   * @return true if there is a solution, false otherwise. 
```

Comment:
```suggestion
   * @param candidateBoard Environment board description.
   * @param nRows Height of the environment board.
   * @param nCols Width of the environment board.
   * @return true if there is a solution, false otherwise. 
```

Some minor style issues, also, should we call the parameter `height` and `width` instead of `nRows` and `nCols`?

Reply:
Sure! I have also thought about it! Height and width do give the user a sense of a gameboard.